File: Linked List/109. Convert Sorted List to Binary Search Tree/109. Convert Sorted List to Binary Search Tree.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

//较为简单的一道题目：可以直接使用递归求解，递归求解的思路：对于给定的链表，首先先查找到链表的中间结点，然后以中间结点作为根节点，左子树由中间结点之前的链表递归生成，右子树由中间结点之后的链表递归生成

class Solution {
public:
    TreeNode* sortedListToBST(ListNode* head) {
        // 1. 处理特殊的空结点情况
        if(!head)             
        return NULL;

        // 2. 处理特殊的叶子结点的情况
        if(head&&!head->next)
        {
            TreeNode *root=new TreeNode(head->val);
            return root;
        }

        // 3. 处理一般链表的情况
        // 3.1 使用快慢双指针法查找到链表的中间结点
        ListNode *fast=head;        //快指针
        ListNode *slow=head;        //慢指针
        ListNode *lastslow=NULL;    //慢指针的前驱结点

        while(fast&&fast->next)
        {
            lastslow=slow;

            fast=fast->next->next;
            slow=slow->next;
        }

        // 3.2 创建中间结点的根节点
        TreeNode *root=new TreeNode(slow->val);

        lastslow->next=NULL;   //断开前半部分链表

        // 3.3 递归生成左子树和右子树
        root->left=sortedListToBST(head);
        root->right=sortedListToBST(slow->next);
        
        return root;
    }
};

File: Linked List/109. Convert Sorted List to Binary Search Tree/109. Convert Sorted List to Binary Search Tree.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def sortedListToBST(self, head):
        """
        :type head: ListNode
        :rtype: TreeNode
        """
        
        # 1. 处理叶子结点和空结点的情况
        if not head:
            return None
        elif head and not head.next:
            root=TreeNode(head.val)
            return root

        # 2. 快慢双指针法找到链表的中间结点
        fast=head    # 快指针
        slow=head    # 慢指针
        lastslow=None  # 慢指针的前驱结点

        while fast and fast.next:
            lastslow=slow

            fast=fast.next.next
            slow=slow.next

        # 3. 根据链表的中间结点生成树的根节点
        root=TreeNode(slow.val)
        
        # 4. 根据前半段链表和后半段链表生成根节点的左子树和右子树
        lastslow.next=None    # 前半段链表封口

        root.left=self.sortedListToBST(head)
        root.right=self.sortedListToBST(slow.next)
        
        return root
        